### PR TITLE
GAWB-311 Move Workspace Access Group deletion to WorkspaceService

### PR DESCRIPTION
**Pull request by jmthibault79**
Delete Intersection Groups as well
- [x] **Submitter**: Rebase to develop. DO NOT SQUASH
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: Make sure documentation for code is complete
- [x] **Submitter**: Review code comments; remove done TODOs, create stories for remaining TODOs
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Add description or comments on the PR explaining the hows/whys (if not obvious)
- [x] Tell  that the PR exists if he wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
- [x] **LR**: Initial review by LR and others.
- [x] Comment / review / update cycle:
  - Rest of team may comments on PR at will
  - **LR assigns to submitter** for feedback fixes
  - Submitter updates documentation as needed
  - Submitter rebases to develop again if necessary
  - Submitter makes further commits. DO NOT SQUASH. **Reassign to LR** for further feedback
- [x]  sign off
- [x] **LR** sign off
- [x] **Assign to submitter** to finalize
- [x] **Submitter**: Squash commits, rebase if necessary
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Ask QA if it's a convenient time to merge this PR**; hold off if they're in the middle of testing something
- [x] **Submitter**: Merge to develop 
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: Check configuration files in Jenkins in case they need changes
- [x] **Submitter**: Verify swagger UI on dev server still works after deployment
- [x] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [x] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 1]
```diff
@@ -201,11 +202,14 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: DataSource,
...
 
-        val accessGroups = workspaceContext.workspace.accessLevels.values.map(containerDAO.authDAO.loadGroup(_, txn)).collect { case Some(g) => g }
-        gcsDAO.deleteWorkspace(workspaceContext.workspace.bucketName, accessGroups.toSeq, bucketDeletionMonitor).map { _ =>
-          containerDAO.authDAO.deleteWorkspaceAccessGroups(workspaceContext.workspace, txn)
-          containerDAO.workspaceDAO.delete(workspaceName, txn)
+        // use a set because the two maps may be duplicates
+        val groupRefs: Set[RawlsGroupRef] = workspaceContext.workspace.accessLevels.values.toSet ++ workspaceContext.workspace.realmACLs.values
+        val groups = groupRefs.map(containerDAO.authDAO.loadGroup(_, txn)).collect { case Some(g) => g }
 
+        Future.traverse(groups) { gcsDAO.deleteGoogleGroup } map { _ =>
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:209]: Formerly GoogleServicesDAO.deleteWorkspace

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 2]
```diff
@@ -201,11 +202,14 @@ class WorkspaceService(protected val userInfo: UserInfo, dataSource: DataSource,
...
+          groupRefs foreach {  containerDAO.authDAO.deleteGroup(_, txn) }
```
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:210]: Formerly AuthDAO.deleteWorkspaceAccessGroups

File: src/test/scala/org/broadinstitute/dsde/rawls/webservice/AdminApiServiceSpec.scala [hunk 3]
```diff
@@ -403,8 +403,8 @@ class AdminApiServiceSpec extends ApiServiceSpec {
     Get("/admin/users") ~>
       sealRoute(services.adminRoutes) ~>
       check {
-        assertResult(RawlsUserInfoList(Seq(userOwner, userWriter, userReader)), response) {
-          responseAs[RawlsUserInfoList]
+        assertResult(Set(userOwner, userWriter, userReader), response) {
```
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/webservice/AdminApiServiceSpec.scala:406]: The order of the results changed.  I don't know why, but the order shouldn't matter, so...

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 1 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:209]: You've used map here and foreach below. Any reason why? Do we go with the rule of "use map when you return results and foreach when you don't"?

File: src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphWorkspaceDAOSpec.scala [hunk 4]
```diff
@@ -40,7 +40,7 @@ class GraphWorkspaceDAOSpec extends FlatSpec with Matchers with OrientDbTestFixt
     }
   }
 
-  it should "save a new workspace with a realm" in withDefaultTestDatabase { dataSource =>
+  it should "save a new workspace with a realm" in withEmptyTestDatabase { dataSource =>
```
Comment (helgridly) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphWorkspaceDAOSpec.scala:43]: I assume just for efficiency?

**Comment by helgridly** [comment]
I have questions, but they're just questions. Overall :+1: , but I'd like  to review this too because I'm not an expert in this jurisdiction.

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 1 above)
Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:209]: That's the rule I'm following, yes

File: src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphWorkspaceDAOSpec.scala (hunk 4 above)
Comment (jmthibault79) [inline src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/GraphWorkspaceDAOSpec.scala:43]: It doesn't depend on existing DB, so this eliminates multiple-workspace conflicts.  I just added another while I'm thinking about it.

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 1 above)
Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:209]: Then you should be using foreach here and not map?

Comment (jmthibault79) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala:209]: No, this returns the RequestComplete

**Comment by dvoet** [comment]
:+1:

**Pull request merged**